Fix do all redirects

jklymak opened the pull request:
(Update 31 Jan 2021):

closes: https://github.com/matplotlib/matplotlib/issues/12374
closes: #25 

Obviously this blows git hub up, but the script that does this is in the first commit...

## Problem 1:  

Currently the top level of the website has a copy of every file that has existed on our webpage, even if the file is obsolete, and not part of current matplotlib docs.  For instance the `/examples/` directory was removed after 2.0.2 (and replaced by `/gallery/`) but is still accessible at https://matplotlib.org/examples/.  @tacaswell wants this to remain so old links do not die, but it also means that search engines think this is a perfectly acceptable current set of webpages, whereas we would like these versions to not show up in searches. 

### Proposed solution:

The script here either soft links all top-level files to their *newest* version in the docs, or makes an html-refresh to do that. 

So, for example `gallery/api` was moved for 3.0.0, so: `ls -halt gallery/api` gives:

```
-rw-r--r--  1 jklymak staff  463 Jan 20 22:24 quad_bezier.html
lrwxr-xr-x  1 jklymak staff   33 Jan 20 22:24 legend.py -> ../../2.2.5/gallery/api/legend.py
-rw-r--r--  1 jklymak staff  463 Jan 20 22:24 radar_chart.html
-rw-r--r--  1 jklymak staff  448 Jan 20 22:24 logos2.html
lrwxr-xr-x  1 jklymak staff   34 Jan 20 22:24 legend.png -> ../../2.2.5/gallery/api/legend.png
```

and `less quad_bezier.html` gives 

```
<!DOCTYPE HTML>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="refresh" content="0;url=https://matplotlib.org/2.2.5/gallery/api/quad_bezier.html" />
        <link rel="canonical" href="https://matplotlib.org/2.2.5/gallery/api/quad_bezier.html" />
    </head>
    <body>
        <h1>
            The page been moved to <a href="https://matplotlib.org/2.2.5/gallery/api/quad_bezier.html"</a>
        </h1>
    </body>
</html>
```

## Problem 2:

Similarly our canonical links go to the top level or the level they were introduced in.  So `https://matplotlib.org/stable/gallery/showcase/mandelbrot.html` has `<link rel="canonical" href="https://matplotlib.org/3.3.4/gallery/showcase/mandelbrot.html"/>` as its canonical link.  Older versions of the docs would link to `<link rel="canonical" href="https://matplotlib.org/gallery/showcase/mandelbrot.html"/>`


### Solution:

The script goes through each html file in all versions (including old versions) and changes the canonical link to the newest version.  So for `quad_bezier.html`:

`less 2.2.5/gallery/api/quad_bezier.html` gives `<link rel="canonical" href="https://matplotlib.org/2.2.5/gallery/api/quad_bezier.html" />`

`less 2.2.4/gallery/api/quad_bezier.html` gives the same link (because 2.2.5 is the newest).

For files that exist in `stable`: 

`less 2.2.4/tutorials/intermediate/artists.html` gives the canonical version in stable. 
`<link rel="canonical" href="https://matplotlib.org/stable/tutorials/intermediate/artists.html" />`

## Maintenance Burden

- the script will need updating if a new top-level subdirectory is added and we don't want it included in the linking
- the script needs to be run when a new version of the docs is released; the script is slow, but I added multi-processing to the canonical links part to move it along.

jklymak commented:
### Procedure for a release:

As before

```
rsync -a 2.0.0/* ./
rm stable
ln -s 2.0.0 stable 
python make_redirects_links.py
```

If a file is removed from 2.0.0, then the make_redirects_links will link it to 1.9.9 (or whatever the last version was).  A new file will then just be linked back to `stable`, which is a bit silly, but it keeps the top level consistent.  

Conversely, we could leave out the rsync step from now on, so new docs are never installed at the top level (newest is always in stable)

jklymak commented:
So I think this should also go back through all the html files, and change the canonical.  canonical would be the latest available version, usually `stable` if the page still exists.  If the page doesn't exist any longer, canonical would be the newest version.

jklymak commented:
A possible improvement of this script might be to put a banner after `<body>` on every old version of the webpages so they are marked as not current.  Of course the most recent page would not get this, and we'd have to make sure we do not add it twice

QuLogic commented on _websiteutils/make_redirects_links.py:
```diff
@@ -0,0 +1,202 @@
+    p = pathlib.Path(fname)
+    for t in tocheck:
+        pnew = pathlib.Path(t, *p.parts[:])
+        if os.path.exists(pnew ):
```
```suggestion
        if pnew.exists():
```

QuLogic commented on _websiteutils/make_redirects_links.py:
```diff
@@ -0,0 +1,202 @@
+    """
+    p = pathlib.Path(fname)
+    for t in tocheck:
+        pnew = pathlib.Path(t, *p.parts[:])
```
Can you not do `Path(t, p)`?

QuLogic commented on _websiteutils/make_redirects_links.py:
```diff
@@ -0,0 +1,202 @@
+            last = findlast(fullname, tocheck)
+            _log.debug(f'Checking: {fullname} found {last}')
+            if last is not None:
+                os.remove('./'+fullname)
```
Adding `./` seems redundant?

QuLogic commented on _websiteutils/make_redirects_links.py:
```diff
@@ -0,0 +1,202 @@
+                    _log.debug(f'new {line}->{ll}')
+                    fout.write(bytes(ll, encoding='utf-8'))
+                else:
+                    fout.write(bytes(line, encoding = 'utf-8'))
```
Why not open `fin` in binary mode as well? Then only `newcanon` needs to be encoded.

jklymak replied:
I did this because you asked, but I'm not sure its an improvement.  No doubt I am missing some tricks to dealing with byte strings....

QuLogic replied:
Ah, I forgot there's no `bytes` f-string equivalent; maybe encoding `new` instead would be better?

jklymak replied:
Yep that works nicer thanks!

jklymak commented:
Note I've dropped the second commit because there is no reason to upload it here until its ready to go.  Let me know if you'd like me to regenerate it, or if one of you would like to do it.

tacaswell commented:
Ah, I have been indpendently working on the script, have some ways to make it faster.

I think it is possible to make the re-directs relative as a kindness to anyone who wants to host these files locally / on an airgapped network.

jklymak commented:
Does a relative redirect work so that the new address looks correct?  We don't want `https://matplotlib.org/boo/who/../../stable/boo/who/index.html`!

tacaswell commented:
I pushed my work, the re-directs now runs in a few seconds on my machine.

Yes, the `../` get eaten by the browser so you get nice looking urls.

jklymak commented:
I admit that I'm not clear how the cache helps the relinking case since the relinked traverses all the files anyway.  

I can see how it helps the canonical case though since the filenames are often repeated.

tacaswell commented:
I think the biggest speed up was dropping the recursive call to `do_links` because `os.walk` already goes through all of the files so we were doing the top-level folders many extra times.  I think we also got a speed up by only sorting out the newest version of a given file once.  That said, I did not profile anything so take that with a grain of salt.

jklymak commented:
Ooops!  Apparently I didn't understand how os. walk worked? Assuming you hit all the files this way...

tacaswell commented:
```
✔ 16:31:39 $ git diff --shortstat
 75899 files changed, 109462 insertions(+), 3506569 deletions(-)
```
The re-direct pages now look like:

```html

<!DOCTYPE HTML>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="refresh" content="0;url=../2.0.2/users/annotations.html" />
        <link rel="canonical" href="https://matplotlib.org/2.0.2/users/annotations.html" />
    </head>
    <body>
        <h1>
            The page been moved <a href="../2.0.2/users/annotations.html">here</a>!
        </h1>
    </body>
</html>

```

Header of pages now looks like

```html

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Matplotlib: Python plotting &#8212; Matplotlib 3.3.2 documentation</title>
    <link rel="stylesheet" href="_static/mpl.css?3.3.2" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/graphviz.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/gallery.css" />
    <link rel="stylesheet" type="text/css" href="_static/gallery-binder.css" />
    <link rel="stylesheet" type="text/css" href="_static/gallery-dataframe.css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="search" type="application/opensearchdescription+xml"
          title="Search within Matplotlib 3.3.2 documentation"
          href="_static/opensearch.xml"/>
    <link rel="shortcut icon" href="_static/favicon.ico"/>

    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
<link rel="top" title="Matplotlib 3.3.2 documentation" href="#" />
    <link rel="canonical" href="https://matplotlib.org/stable/index.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>

```

jklymak commented:
That looks great.  Thanks @tacaswell.

While doing this, do we want to put a warning block on all the outdated versions of the webpage?  i.e. https://github.com/matplotlib/matplotlib.github.com/pull/25 

ie. add to body:

if exists in stable:
```
<body><div id="olddocs-message"> 
             You are reading an old version of the documentation (3.3.2).  For the latest version see <a href="/stable/tutorials/introductory/usage.html">https://matplotlib.org/stable/tutorials/introductory/usage.html</a>
 </div>
```

if it doesn't exist in stable:
```
<body><div id="olddocs-message"> 
             You are reading an old version of the documentation (3.3.2).  For the latest  docs see <a href="/stable/">https://matplotlib.org/stable/</a>
 </div>
```

jklymak commented:
Added the banner logic.  

It somewhat fragilely assumes that the `<body>` tag is on a line of its own.  If we wanted to use BeautifulSoup and were willing to prettify the output, we could do this much more robustly.  But I wasn't sure if that was too invasive.  

Also removed the double recursion under do_canonical!  Its quite fast now and I'm 90% sure it hits everything.

jklymak commented:
Note I don't think this needs to wait for https://github.com/matplotlib/matplotlib/pull/19456

jklymak commented:
This is working so far as I can tell.  Header and start of body now look like:

```html
...
    <link rel="canonical" href="https://matplotlib.org/stable/index.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
<div id="olddocs-message"> You are reading an old version of thedocumentation (v3.3.2).  For the latest version see <a href="https://matplotlib.org/stable/index.html">https://matplotlib.org/stable/index.html</a></div>


```

jklymak commented:
@tacaswell @QuLogic I don't see any reason to not move forward with this.  If you do, happy to chat, but if its OK, I think implementing it sooner rather than later is preferable...

QuLogic commented on _websiteutils/make_redirects_links.py:
```diff
@@ -145,16 +140,12 @@ def do_canonicals(dname):
-                last = findlast(basename, tocheck)
-                if last is not None:
-                    update_canonical(fullname, last, dname == tocheck[1])
+    for fullname in dname.rglob("*.html"):
```
Note, `Path.rglob` doesn't seem to support multiple patterns, but we do not have any `.htm` files.

QuLogic commented:
I pushed a few cleanups and improvements.

jklymak commented:
Hmmm, is `functools.cache` 3.9 only?

QuLogic commented:
Oh, yes, but it's basically a lighter version of `lru_cache`; we can switch to that if you prefer.

jklymak commented:
I don't mind, I just need to upgrade my env

jklymak commented on _websiteutils/make_redirects_links.py:
```diff
@@ -0,0 +1,242 @@
+    if not args.no_canonicals:
+        if np is not None:
+            with multiprocessing.Pool(np) as pool:
+                pool.map(do_canonicals, tocheck[1:])
```
This option now fails: 
```
Traceback (most recent call last):
  File "/Users/jklymak/anaconda3/envs/matplotlibdev/lib/python3.9/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/Users/jklymak/anaconda3/envs/matplotlibdev/lib/python3.9/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/Users/jklymak/matplotlib.github.com/_websiteutils/make_redirects_links.py", line 142, in do_canonicals
    last = findlast(basename, tocheck)
TypeError: unhashable type: 'list'
```

jklymak replied:
Perhaps just remove it?

QuLogic replied:
Ah, `map` must convert it from a tuple to a list.

jklymak replied:
so  its the cache that is causing the problem?  Happy to remove my optimization in favour of your optimization ;-)

QuLogic replied:
Actually no, this works fine for me; do you have some stashed changes? `tocheck` should be a tuple after df8ed616dd53bcf4393d3c10b672eb622d0d468d (which was before 6fcc3b744733fb9081825a96731fa5d3b42ce812).

jklymak replied:
That aside, I think this is working great....

jklymak commented:
Changed the banner tag div id from "olddocs-message" which doesn't exist, to "unreleased-message" which while not quite accurate, already exists in many of the old versions.  Gives a banner that is sticky at the top of the viewport:

<img width="1207" alt="Banner" src="https://user-images.githubusercontent.com/1562854/107459780-bf1b9400-6b0b-11eb-8a72-64360acf659e.png">
